src.ts/transaction/quai-transaction.ts

```diff
@@ -142,12 +128,21 @@ export class QuaiTransaction extends AbstractTransaction<Signature> implements Q
             throw new Error("Cross-shard & cross-ledger transactions are not supported");
         }
 
-        let hash = keccak256(data)
+        let hash = keccak256(this.serialized)
         hash = '0x' + this.originShard+ (originUtxo ? 'F' : '1') + hash.charAt(5) + this.originShard+ (destUtxo ? 'F' : '1') + hash.slice(9)
```

Comment:
The implementation for getting the permuted QIP10 hash of a transaction should, for all intents and purposes, match that of `go-quai`. This means that we should be manipulating the hash as a uint8[] rather than string and setting the values that correspond to the first 4 elements of the uint8[]. Here is the implementation for how `go-quai` handles computing a Quai tx hash:


```go
hash = crypto.Keccak256Hash(data)
// the origin of this tx is the *destination* of the utxos being spent
if len(location) == 2 {
  origin := (uint8(location[0]) * 16) + uint8(location[1])
  hash[0] = origin
  hash[1] &= 0x7F // 01111111 in binary (set first bit to 0)
  hash[2] = origin
  hash[3] &= 0x7F
} else {
  from, err := Sender(NewSigner(tx.ChainId(), common.Location{0, 0}), tx) // location not important when performing ecrecover
  if err != nil {
    return h // Caller of this function will fail with wrong tx hash and will appropriately handle the error
  }
  location := *from.Location()
  origin := (uint8(location[0]) * 16) + uint8(location[1])
  hash[0] = origin
  hash[1] &= 0x7F
  hash[2] = origin
  hash[3] &= 0x7F
}
```